<html><head>
<META http-equiv="Content-Type" content="text/html">
<!--SafClassName="docSection1Title"--><!--SafTocEntry="6.4 Exception Objects"-->
<link rel="STYLESHEET" type="text/css" href="FILES/style.css">
<link rel="STYLESHEET" type="text/css" href="FILES/docsafari.css">
<style type="text/css">	.tt1    {font-size: 10pt;}</style>
</head>
<body>
<table width="100%" border="0" cellspacing="0" cellpadding="0">
<td class="tt1"><a href="NFO/lib.html">[ Team LiB ]</a></td><td valign="top" class="tt1" align="right">
	<a href="0596001886_pythonian-chp-6-sect-3.html"><img src="FILES/previous.gif" width="62" height="15" border="0" align="absmiddle" alt="Previous Section"></a>
	<a href="0596001886_pythonian-chp-6-sect-5.html"><img src="FILES/next.gif" width="41" height="15" border="0" align="absmiddle" alt="Next Section"></a>
</td></table>
<br>
<table width="100%" border="0" cellspacing="0" cellpadding="0">
<tr>
<td valign="top"><A NAME="pythonian-CHP-6-SECT-4"></A>
<H3 class="docSection1Title">6.4 Exception Objects</H3>

<P class="docText"><A NAME="pythonian-CHP-6-ITERM-4572"></A> <A NAME="pythonian-CHP-6-ITERM-4573"></A>Exceptions are instances of
subclasses of the built-in <TT>Exception</TT> class. For
backward compatibility, Python also lets you use strings, or
instances of any class, as exception objects, but such usage risks
future incompatibility and gives no benefits. An instance of any
subclass of <TT>Exception</TT> has an attribute
<TT>args</TT>, the tuple of arguments used to create the
instance. <TT>args</TT> holds error-specific information,
usable for diagnostic or recovery
purposes.<A NAME="pythonian-CHP-6-ITERM-4574"></A> <A NAME="pythonian-CHP-6-ITERM-4575"></A>
<A NAME="pythonian-CHP-6-ITERM-4576"></A> <A NAME="pythonian-CHP-6-ITERM-4577"></A></P>

<A NAME="pythonian-CHP-6-SECT-4.1"></A>
<H4 class="docSection2Title">6.4.1 The Hierarchy of Standard Exceptions</H4>

<P class="docText"><A NAME="pythonian-CHP-6-ITERM-4578"></A> <A NAME="pythonian-CHP-6-ITERM-4579"></A>All exceptions that Python
itself raises are instances of subclasses of
<TT>Exception</TT>. The inheritance structure of exception
classes is important, as it determines which
<TT>except</TT> clauses handle which exceptions.</P>

<P class="docText">The <TT>SystemExit</TT> class inherits directly from
<TT>Exception</TT>. Instances of
<TT>SystemExit</TT> are normally raised by the
<TT>exit</TT> function in module <TT>sys</TT>
(covered in <A class="docLink" HREF="0596001886_pythonian-chp-8.html#pythonian-CHP-8">Chapter 8</A>).<A NAME="pythonian-CHP-6-ITERM-4580"></A> <A NAME="pythonian-CHP-6-ITERM-4581"></A> <A NAME="pythonian-CHP-6-ITERM-4582"></A></P>

<P class="docText"><A NAME="pythonian-CHP-6-ITERM-4583"></A>
<A NAME="pythonian-CHP-6-ITERM-4584"></A>
<A NAME="pythonian-CHP-6-ITERM-4585"></A>
<A NAME="pythonian-CHP-6-ITERM-4586"></A>Other
standard exceptions derive from <TT>StandardError</TT>, a
direct subclass of <TT>Exception</TT>. Three subclasses of
<TT>StandardError</TT>, like
<TT>StandardError</TT> itself and
<TT>Exception</TT>, are never instantiated directly. Their
purpose is to make it easier for you to specify
<TT>except</TT> clauses that handle a broad range of
related errors. These subclasses are:</P>

<A NAME="pythonian-CHP-6-ITERM-4587"></A><A NAME="pythonian-CHP-6-ITERM-4588"></A><A NAME="pythonian-CHP-6-ITERM-4589"></A><A NAME="pythonian-CHP-6-ITERM-4590"></A><DL class="docList">
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">ArithmeticError</span></span></span></DT>
<DD>
<P class="docList">The base class for exceptions due to arithmetic errors (i.e.,
<TT>OverflowError</TT>,
<TT>ZeroDivisionError</TT>,
<TT>FloatingPointError</TT>)</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">LookupError</span></span></span></DT>
<DD>
<P class="docList">The base class for exceptions that a container raises when it
receives an invalid key or index (i.e.,
<TT>IndexError</TT>, <TT>KeyError</TT>)</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">EnvironmentError</span></span><A NAME="pythonian-CHP-6-ITERM-4587"></A>
<A NAME="pythonian-CHP-6-ITERM-4588"></A>
<A NAME="pythonian-CHP-6-ITERM-4589"></A>
<A NAME="pythonian-CHP-6-ITERM-4590"></A></span></DT>
<DD>
<P class="docList">The base class for exceptions due to external causes (i.e.,
<TT>IOError</TT>, <TT>OSError</TT>,
<TT>WindowsError</TT>)</P>
</DD>
</DL>


<A NAME="pythonian-CHP-6-SECT-4.2"></A>
<H4 class="docSection2Title">6.4.2 Standard Exception Classes</H4>

<P class="docText"><A NAME="pythonian-CHP-6-ITERM-4591"></A> <A NAME="pythonian-CHP-6-ITERM-4592"></A>Common
runtime errors raise exceptions of the following classes:</P>

<A NAME="pythonian-CHP-6-ITERM-4593"></A><A NAME="pythonian-CHP-6-ITERM-4594"></A><A NAME="pythonian-CHP-6-ITERM-4595"></A><A NAME="pythonian-CHP-6-ITERM-4596"></A><A NAME="pythonian-CHP-6-ITERM-4597"></A><A NAME="pythonian-CHP-6-ITERM-4598"></A><A NAME="pythonian-CHP-6-ITERM-4599"></A><A NAME="pythonian-CHP-6-ITERM-4600"></A><A NAME="pythonian-CHP-6-ITERM-4601"></A><A NAME="pythonian-CHP-6-ITERM-4602"></A><A NAME="pythonian-CHP-6-ITERM-4603"></A><A NAME="pythonian-CHP-6-ITERM-4604"></A><A NAME="pythonian-CHP-6-ITERM-4605"></A><A NAME="pythonian-CHP-6-ITERM-4606"></A><A NAME="pythonian-CHP-6-ITERM-4607"></A><A NAME="pythonian-CHP-6-ITERM-4608"></A><A NAME="pythonian-CHP-6-ITERM-4609"></A><A NAME="pythonian-CHP-6-ITERM-4610"></A><A NAME="pythonian-CHP-6-ITERM-4611"></A><A NAME="pythonian-CHP-6-ITERM-4612"></A><A NAME="pythonian-CHP-6-ITERM-4613"></A><A NAME="pythonian-CHP-6-ITERM-4614"></A><A NAME="pythonian-CHP-6-ITERM-4615"></A><A NAME="pythonian-CHP-6-ITERM-4616"></A><A NAME="pythonian-CHP-6-ITERM-4617"></A><A NAME="pythonian-CHP-6-ITERM-4618"></A><A NAME="pythonian-CHP-6-ITERM-4619"></A><A NAME="pythonian-CHP-6-ITERM-4620"></A><A NAME="pythonian-CHP-6-ITERM-4621"></A><A NAME="pythonian-CHP-6-ITERM-4622"></A><A NAME="pythonian-CHP-6-ITERM-4623"></A><A NAME="pythonian-CHP-6-ITERM-4624"></A><A NAME="pythonian-CHP-6-ITERM-4625"></A><A NAME="pythonian-CHP-6-ITERM-4626"></A><A NAME="pythonian-CHP-6-ITERM-4627"></A><A NAME="pythonian-CHP-6-ITERM-4628"></A><A NAME="pythonian-CHP-6-ITERM-4629"></A><A NAME="pythonian-CHP-6-ITERM-4630"></A><A NAME="pythonian-CHP-6-ITERM-4631"></A><A NAME="pythonian-CHP-6-ITERM-4632"></A><A NAME="pythonian-CHP-6-ITERM-4633"></A><A NAME="pythonian-CHP-6-ITERM-4634"></A><A NAME="pythonian-CHP-6-ITERM-4635"></A><A NAME="pythonian-CHP-6-ITERM-4636"></A><A NAME="pythonian-CHP-6-ITERM-4637"></A><A NAME="pythonian-CHP-6-ITERM-4638"></A><A NAME="pythonian-CHP-6-ITERM-4639"></A><A NAME="pythonian-CHP-6-ITERM-4640"></A><A NAME="pythonian-CHP-6-ITERM-4641"></A><A NAME="pythonian-CHP-6-ITERM-4642"></A><A NAME="pythonian-CHP-6-ITERM-4643"></A><A NAME="pythonian-CHP-6-ITERM-4644"></A><A NAME="pythonian-CHP-6-ITERM-4645"></A><A NAME="pythonian-CHP-6-ITERM-4646"></A><A NAME="pythonian-CHP-6-ITERM-4647"></A><A NAME="pythonian-CHP-6-ITERM-4648"></A><DL class="docList">
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">AssertionError</span></span><A NAME="pythonian-CHP-6-ITERM-4593"></A>
<A NAME="pythonian-CHP-6-ITERM-4594"></A></span></DT>
<DD>
<P class="docList">An assert statement failed.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">AttributeError</span></span><A NAME="pythonian-CHP-6-ITERM-4595"></A>
<A NAME="pythonian-CHP-6-ITERM-4596"></A>
<A NAME="pythonian-CHP-6-ITERM-4597"></A></span></DT>
<DD>
<P class="docList">An attribute reference or assignment failed.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">FloatingPointError</span></span><A NAME="pythonian-CHP-6-ITERM-4598"></A>
<A NAME="pythonian-CHP-6-ITERM-4599"></A></span></DT>
<DD>
<P class="docList">A floating-point operation failed. Derived from
<TT>ArithmeticError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">IOError</span></span><A NAME="pythonian-CHP-6-ITERM-4600"></A>
<A NAME="pythonian-CHP-6-ITERM-4601"></A>
<A NAME="pythonian-CHP-6-ITERM-4602"></A></span></DT>
<DD>
<P class="docList">An I/O operation failed (e.g., the disk is full, a file was not
found, or needed permissions were missing). Derived from
<TT>EnvironmentError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">ImportError</span></span><A NAME="pythonian-CHP-6-ITERM-4603"></A>
<A NAME="pythonian-CHP-6-ITERM-4604"></A></span></DT>
<DD>
<P class="docList">An <TT>import</TT> statement (covered in <A class="docLink" HREF="0596001886_pythonian-chp-7.html#pythonian-CHP-7">Chapter 7</A>) cannot find the module to import or cannot
find a name specifically requested from the module.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">IndentationError</span></span><A NAME="pythonian-CHP-6-ITERM-4605"></A>
<A NAME="pythonian-CHP-6-ITERM-4606"></A>
<A NAME="pythonian-CHP-6-ITERM-4607"></A></span></DT>
<DD>
<P class="docList">The parser encountered a syntax error due to incorrect indentation.
Derived from <TT>SyntaxError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">IndexError</span></span><A NAME="pythonian-CHP-6-ITERM-4608"></A>
<A NAME="pythonian-CHP-6-ITERM-4609"></A>
<A NAME="pythonian-CHP-6-ITERM-4610"></A></span></DT>
<DD>
<P class="docList">An integer used to index a sequence is out of range (using a
non-integer as a sequence index raises <TT>TypeError</TT>).
Derived from <TT>LookupError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">KeyError</span></span><A NAME="pythonian-CHP-6-ITERM-4611"></A>
<A NAME="pythonian-CHP-6-ITERM-4612"></A>
<A NAME="pythonian-CHP-6-ITERM-4613"></A></span></DT>
<DD>
<P class="docList">A key used to index a mapping is not in the mapping. Derived from
<TT>LookupError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">KeyboardInterrupt</span></span><A NAME="pythonian-CHP-6-ITERM-4614"></A>
<A NAME="pythonian-CHP-6-ITERM-4615"></A></span></DT>
<DD>
<P class="docList">The user pressed the interrupt key (Ctrl-C, Ctrl-Break, or Delete,
depending on the platform).</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">MemoryError</span></span><A NAME="pythonian-CHP-6-ITERM-4616"></A>
<A NAME="pythonian-CHP-6-ITERM-4617"></A>
<A NAME="pythonian-CHP-6-ITERM-4618"></A></span></DT>
<DD>
<P class="docList">An operation ran out of memory.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">NameError</span></span><A NAME="pythonian-CHP-6-ITERM-4619"></A>
<A NAME="pythonian-CHP-6-ITERM-4620"></A>
<A NAME="pythonian-CHP-6-ITERM-4621"></A></span></DT>
<DD>
<P class="docList">A variable was referenced, but its name is not bound.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">NotImplementedError</span></span><A NAME="pythonian-CHP-6-ITERM-4622"></A>
<A NAME="pythonian-CHP-6-ITERM-4623"></A></span></DT>
<DD>
<P class="docList">Raised by abstract base classes to indicate that a concrete subclass
must override a method.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">OSError</span></span><A NAME="pythonian-CHP-6-ITERM-4624"></A>
<A NAME="pythonian-CHP-6-ITERM-4625"></A>
<A NAME="pythonian-CHP-6-ITERM-4626"></A></span></DT>
<DD>
<P class="docList">Raised by functions in module <TT>os</TT> (covered in
<A class="docLink" HREF="0596001886_pythonian-chp-10.html#pythonian-CHP-10">Chapter 10</A> and <A class="docLink" HREF="0596001886_pythonian-chp-14.html#pythonian-CHP-14">Chapter 14</A>) to indicate platform-dependent errors.
Derived from <TT>EnvironmentError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">OverflowError</span></span><A NAME="pythonian-CHP-6-ITERM-4627"></A>
<A NAME="pythonian-CHP-6-ITERM-4628"></A></span></DT>
<DD>
<P class="docList">The result of an operation on an integer is too large to fit into an
integer (operator <TT>&lt;&lt;</TT> does not raise this
exception: rather, it drops excess bits). Derived from
<TT>ArithmeticError</TT>. Python 2.1 only; in 2.2 and 2.3,
too-large integer results implicitly become long integers, without
raising exceptions.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">SyntaxError</span></span><A NAME="pythonian-CHP-6-ITERM-4629"></A>
<A NAME="pythonian-CHP-6-ITERM-4630"></A></span></DT>
<DD>
<P class="docList">The parser encountered a syntax error.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">SystemError</span></span><A NAME="pythonian-CHP-6-ITERM-4631"></A>
<A NAME="pythonian-CHP-6-ITERM-4632"></A>
<A NAME="pythonian-CHP-6-ITERM-4633"></A></span></DT>
<DD>
<P class="docList">An internal error within Python itself or some extension module. You
should report this to the authors and maintainers of Python, or of
the extension in question, with all possible details to allow
reproducing it.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">TypeError</span></span><A NAME="pythonian-CHP-6-ITERM-4634"></A>
<A NAME="pythonian-CHP-6-ITERM-4635"></A></span></DT>
<DD>
<P class="docList">An operation or function was applied to an object of an inappropriate
type.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">UnboundLocalError</span></span><A NAME="pythonian-CHP-6-ITERM-4636"></A>
<A NAME="pythonian-CHP-6-ITERM-4637"></A></span></DT>
<DD>
<P class="docList">A reference was made to a local variable, but no value is currently
bound to that local variable. Derived from
<TT>NameError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">UnicodeError</span></span><A NAME="pythonian-CHP-6-ITERM-4638"></A>
<A NAME="pythonian-CHP-6-ITERM-4639"></A></span></DT>
<DD>
<P class="docList">An error occurred while converting Unicode to a string or vice versa.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">ValueError</span></span><A NAME="pythonian-CHP-6-ITERM-4640"></A>
<A NAME="pythonian-CHP-6-ITERM-4641"></A></span></DT>
<DD>
<P class="docList">An operation or function was applied to an object that has a correct
type but an inappropriate value, and nothing more specific (e.g.,
<TT>KeyError</TT>) applies.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">WindowsError</span></span><A NAME="pythonian-CHP-6-ITERM-4642"></A>
<A NAME="pythonian-CHP-6-ITERM-4643"></A></span></DT>
<DD>
<P class="docList">Raised by functions in module <TT>os</TT> (covered in
<A class="docLink" HREF="0596001886_pythonian-chp-10.html#pythonian-CHP-10">Chapter 10</A> and <A class="docLink" HREF="0596001886_pythonian-chp-14.html#pythonian-CHP-14">Chapter 14</A>) to indicate Windows-specific errors. Derived
from <TT>OsError</TT>.</P>
</DD>
<DT><span class="docPubcolor"><span class="docPubcolor"><span class="docMonofont">ZeroDivisionError</span></span><A NAME="pythonian-CHP-6-ITERM-4644"></A>
<A NAME="pythonian-CHP-6-ITERM-4645"></A></span></DT>
<DD>
<P class="docList">A divisor (the right-hand operand of a <TT>/</TT>,
<TT>//</TT>, or <TT>%</TT> operator or the second
argument to built-in function <TT>divmod</TT>) is
<TT>0</TT>. Derived from
<TT>ArithmeticError</TT>.<A NAME="pythonian-CHP-6-ITERM-4646"></A> <A NAME="pythonian-CHP-6-ITERM-4647"></A> <A NAME="pythonian-CHP-6-ITERM-4648"></A></P>
</DD>
</DL>



<a href="0596001886_"><img src="FILES/pixel.gif" width="1" height="1" border="0"></a><ul></ul>
</td>
</tr>
</table>
<td></td>
<table width="100%" border="0" cellspacing="0" cellpadding="0">
<td class="tt1"><a href="NFO/lib.html">[ Team LiB ]</a></td><td valign="top" class="tt1" align="right">
          <a href="0596001886_pythonian-chp-6-sect-3.html"><img src="FILES/previous.gif" width="62" height="15" border="0" align="absmiddle" alt="Previous Section"></a>
          <a href="0596001886_pythonian-chp-6-sect-5.html"><img src="FILES/next.gif" width="41" height="15" border="0" align="absmiddle" alt="Next Section"></a>
</td></table>
</body></html>
